presentation/src/main/java/org/sagebionetworks/research/presentation/perform_task/PerformTaskViewModel.java

```diff
@@ -211,51 +222,72 @@ public TaskView getTaskView() {
+     * @param taskResult The task result before this switch occured.
+     */
+    protected void updateCurrentStep(@Nullable Step nextStep, @NonNull TaskResult taskResult) {
+        nextStep = resolveSection(nextStep);
+        if (nextStep == null) {
+            this.currentStepLiveData.setValue(null);
+            this.stepViewLiveData.setValue(null);
+            this.taskProgressLiveData.setValue(null);
+        } else {
+            TaskProgress nextProgress = this.stepNavigator.getProgress(nextStep, taskResult);
+            this.taskProgressLiveData.setValue(nextProgress);
+            LOGGER.debug("Setting step: {}", nextStep);
+            this.currentStepLiveData.setValue(nextStep);
+            StepView stepView = this.stepViewMapping.get(nextStep);
+            this.stepViewLiveData.setValue(stepView);
+        }
+    }
 
-        Step backStep = stepNavigator.getPreviousStep(currentStep, taskResult);
-        StepView stepView = null;
-        if (backStep != null) {
-            TaskProgress backProgress = stepNavigator.getProgress(backStep, taskResult);
-            taskProgressLiveData.setValue(backProgress);
-            LOGGER.debug("Setting backStep: {}", backStep);
-            currentStepLiveData.setValue(backStep);
-            stepView = this.stepViewMapping.get(backStep);
-            if (stepView.shouldSkip(taskResult)) {
-                this.goBack();
-                return;
-            }
+    protected Step resolveSection(@Nullable Step step) {
+        while (step instanceof SectionStep) {
+            step = ((SectionStep) step).getSteps().get(0);
         }
 
-        stepViewLiveData.setValue(stepView);
+        return step;
     }
 
+    /**
+     * Navigates forward in the task writing a result for the current step.
+     */
     public void goForward() {
         LOGGER.debug("goForward called");
         Step currentStep = currentStepLiveData.getValue();
         TaskResult taskResult = taskResultLiveData.getValue();
         checkState(taskResult != null);
-
-        Step nextStep = stepNavigator.getNextStep(currentStep, taskResult);
-        StepView stepView = null;
-        if (nextStep != null) {
-            TaskProgress nextProgress = stepNavigator.getProgress(nextStep, taskResult);
-            taskProgressLiveData.setValue(nextProgress);
-            LOGGER.debug("Setting forwardStep: {}", nextStep);
-            currentStepLiveData.setValue(nextStep);
-            stepView = this.stepViewMapping.get(nextStep);
-            if (stepView.shouldSkip(taskResult)) {
-                this.goForward();
-                return;
+        if (currentStep != null) {
+            // We write a step result if we don't already have one
+            Result previousResult = taskResult.getResult(currentStep);
+            if (previousResult == null || previousResult instanceof ResultBase) {
+                this.addStepResult(currentStep.instantiateStepResult());
```

Comment:
This was an attempt to make sure that every step creates at least one result, which has it's identifier. This makes it so the task result can be used to identify if a step has gone which is needed to figure out which hand should go next. I did the whole previous result thing to avoid overwriting any more complicated results that the step chose to write. To the question about a step adding multiple result objects it definitely can however only one of those results is allowed to have the same identifier as the step.